random_numbers.c

```diff
@@ -0,0 +1,154 @@
+
+/* The following code offers an alterate pseudo random number generator
+   namely XORSHIFT128+ to use instead of C's rand(). Its performance is 
+   on par with C's rand().
+ */
+
+
+
+#include <stdint.h>
+#ifdef __GNUC__
+#include <sys/time.h>
+#endif
+#include <time.h>
+#ifdef __linux__
+    #include <sys/syscall.h>
+    #define GRND_NONBLOCK	1
+#endif
+#include "n2n.h"
+#include "random_numbers.h"
+
+struct rn_generator_state_t {
+    uint64_t a, b;
+ };
+
+/* The state must be seeded so that it is not all zero, choose some
+   arbitrary defaults (in this case: taken from splitmix64)
+ */
+static struct rn_generator_state_t rn_current_state
+			       = { .a    = 0x9E3779B97F4A7C15,
+			           .b    = 0xBF58476D1CE4E5B9
+};
+
+static uint64_t reseed_counter = 0;
+
+/* taken from benchmark.c */
+#if defined(WIN32) && !defined(__GNUC__)
+#include <windows.h>
+static int gettimeofday(struct timeval *tp, void *tzp)
+{
+    time_t clock;
+    struct tm tm;
+    SYSTEMTIME wtm;
+    GetLocalTime(&wtm);
+    tm.tm_year = wtm.wYear - 1900;
+    tm.tm_mon = wtm.wMonth - 1;
+    tm.tm_mday = wtm.wDay;
+    tm.tm_hour = wtm.wHour;
+    tm.tm_min = wtm.wMinute;
+    tm.tm_sec = wtm.wSecond;
+    tm.tm_isdst = -1;
+    clock = mktime(&tm);
+    tp->tv_sec = clock;
+    tp->tv_usec = wtm.wMilliseconds * 1000;
+    return (0);
+}
+#endif
+
+static int32_t set_seed (uint64_t seed) {
```

Comment:
This function mixes previous state, time, and user input together (depending on the user input) and this is not good. Why can't you simply use the inizialization described in https://en.wikipedia.org/wiki/Xorshift#Initialization ?

Reply:
We know that rand() is deterministic and returns the same result given a seed. We don't have many expectations on rand() because we know that it is a naive random generator. By mixing different features into a custom random generation I think that we won't know the exact features of such random generator and we might think that it "magical" working great in all the situations and platforms, but this is unlikely to be true since it is a custom generator implemented only in this project. I would just take the standard xorshift+ if it provides substantial advantages over the current rand(), but avoid customized implementations.

Reply:
Done.